### DISCUSSION TITLE
Beginner question about mpc.input and mpc.output

### DISCUSSION THREAD

**Discussion by MargaroneGiada** [lschoe/mpyc#95] [Q&A] [answered]
Hello, 
I am a student  and, as a beginner in MPC field, I have encountered some challenges and would appreciate your assistance in clarifying a few points.

I am trying to understand how MPyC handles secret inputs and outputs. I wrote the following code:
`print(await mpc.output(mpc.input(mpc.SecInt(16)(1))))`
My understanding is that mpc.input should accept a secret value, and mpc.output should reveal this value. However, when I run this code, I can see all secret values, mine and those of the other parties, in clear text. This confuses me, as I thought the values should remain secret during the computation.

Could you explain why this happens and if there is a way to ensure the values remain secret until the final output phase? Additionally, am I misunderstanding something fundamental about how MPyC handles inputs and outputs?

Thank you very much, I appreciate any clarification you can provide.

**Comment by lschoe** [comment, marked as answer]
Not sure what you've been printing. The input values are really split into random secret shares, and then combined again to print the output in the clear. Maybe run the next program to see this in action:
```python
from mpyc.runtime import mpc

async def main():
    async with mpc:
        a = mpc.input(mpc.SecInt(16)(1))  # each party splits 1 into m shares
        a = await mpc.gather(a)  # wait for the secret shares, one per party
        print(a)  # prints list with m random shares
        print(await mpc.output(a))  # prints [1] * m 
                
mpc.run(main())
````
This gives as result, when running it with 5 parties:
```
$ python disc95.py -M5
2024-06-05 23:08:27,757 Start MPyC runtime v0.10.1
2024-06-05 23:08:28,300 All 5 parties connected.
[100704201444869, -110283150495623, 44321259337119, -4615352456634, 111627973283265]
[1, 1, 1, 1, 1]
2024-06-05 23:08:28,311 Stop MPyC -- elapsed time: 0:00:00.006|bytes sent: 156
```

**Comment by aiwushaqi** [comment]
11
